### DISCUSSION TITLE
error in source() when generating report

### DISCUSSION THREAD

**Discussion by abaumane** [IBAHCM/RPiR#313] [Archive] [answered]
Hi @richardreeve @soniamitchell,

I want to source a file in the demo folder at the beginning of another demo:
``` 
source("demo/populations.R")
```
It works perfectly when I run it within R, but when I want to compile a report it spits out this error:
```
Error in file(filename, "r", encoding = encoding) : 
  cannot open the connection
Calls: <Anonymous> ... withCallingHandlers -> withVisible -> eval -> eval -> source -> file

Execution halted
```
All I could find online about this was to do with directories, but my working directory is the project directory, so I don't see how that could be a problem.

Do you have any advice on what to do? Is there another way that I could maybe source that file so that the HTML report compiles?

Thanks :)

**Comment by abaumane** [comment]
I did also try putting it into the main project folder but got the same error.

**Comment by abaumane** [comment]
Update: I tried putting setwd() before the source() command and it finally compiled a report! But it only does it with setwd() in. 

Is this a problem for others accessing the package from other systems? Obviously they wouldn't have the same path.

**Comment by soniamitchell** [comment]
Hi @abaumane, why do you want to do this? I've made a little package that does what you ask and added you as a contributor (https://github.com/IBAHCM/demoSource). But I'm not sure sourcing a file from a demo is a good idea, though I could be wrong! (I've just never needed to do it myself)

**Comment by soniamitchell** [comment]
As a general rule though, you should try not to use `setwd()` in your scripts. 

And to explain what I've done. The contents of the `inst` directory are installed as part of your package (and dumped in the root of your package). The `system.file()` function will locate the whereabouts of installed files (after you `install()` your package).

**Comment by abaumane** [comment]
@soniamitchell I wanted to separate some code so I don't have to copy and paste it in multiple demos, like we did in series 2. That was my attempt of coming up with a 'cleverer' solution for this part in Project 2:

> Your demo will probably need to start with the same block of code that the last demo did, so that it generates the same populations. You can just copy it from one demo to the next (or you may come up with a cleverer way of dealing with this problem).

Is sourcing some code within another file something that's not possible in a package? Or it is, but the file should be in another folder? 

Thank you for making the package! I don't quite understand what it does, but I'll see if I can make sense of it. Worst case scenario, I can just make do with copy and paste.

**Reply by soniamitchell** [reply]
It’s possible, I’m just not sure if it’s best practice :)

**Comment by abaumane** [comment, marked as answer]
Actually @soniamitchell I've just come up with a better method for what I wanted to do, that does not need any of this (and it works!), so I've solved the problem. Sorry for bothering you so much and thanks for trying to help!

**Reply by soniamitchell** [reply]
Fab! Looking forward to seeing what you did :)
